File: tools/src/idevice_id.rs
// Jackson Coxson
// Gets the devices from the muxer

use futures_util::StreamExt;
use idevice::usbmuxd::{UsbmuxdAddr, UsbmuxdConnection};

#[tokio::main]
async fn main() {
    tracing_subscriber::fmt::init();

    let addr = UsbmuxdAddr::from_env_var().expect("Failed to parse USBMUXD_SOCKET_ADDRESS");
    let muxer = addr.to_socket().await.unwrap();
    let mut muxer = UsbmuxdConnection::new(muxer, 0);
    let res = muxer.get_devices().await.unwrap();
    println!("{res:#?}");

    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 && args[1] == "-l" {
        let mut s = muxer.listen().await.expect("listen failed");
        while let Some(dev) = s.next().await {
            let dev = dev.expect("failed to read from stream");
            println!("{dev:#?}");
        }
    }
}

File: cpp/examples/idevice_id.cpp
// Jackson Coxson

#include <idevice++/usbmuxd.hpp>
#include <iostream>

int main() {
    auto u = IdeviceFFI::UsbmuxdConnection::default_new(0).expect("failed to connect to usbmuxd");
    auto devices = u.get_devices().expect("failed to get devices from usbmuxd");

    for (IdeviceFFI::UsbmuxdDevice& d : devices) {
        auto udid = d.get_udid();
        if (udid.is_none()) {
            std::cerr << "failed to get udid";
            continue;
        }
        auto connection_type = d.get_connection_type();
        if (connection_type.is_none()) {
            std::cerr << "failed to get connection type";
            continue;
        }
        std::cout << udid.unwrap() << " (" << connection_type.unwrap().to_string() << ")" << "\n";
    }
}
